src/cpu-kernels/getitem.cpp

```diff
@@ -1082,3 +1082,25 @@ ERROR awkward_listarrayU32_getitem_jagged_descend_64(int64_t* tooffsets, const i
 ERROR awkward_listarray64_getitem_jagged_descend_64(int64_t* tooffsets, const int64_t* slicestarts, int64_t slicestartsoffset, const int64_t* slicestops, int64_t slicestopsoffset, int64_t sliceouterlen, const int64_t* fromstarts, int64_t fromstartsoffset, const int64_t* fromstops, int64_t fromstopsoffset) {
   return awkward_listarray_getitem_jagged_descend<int64_t, int64_t>(tooffsets, slicestarts, slicestartsoffset, slicestops, slicestopsoffset, sliceouterlen, fromstarts, fromstartsoffset, fromstops, fromstopsoffset);
 }
+
+template<typename T>
+EXPORT_SYMBOL T getitem_at_nowrap_cpu(const std::shared_ptr<T> ptr_, size_t at) {
+  return ptr_.get()[at];
+}
+
+template<typename T>
+EXPORT_SYMBOL void setitem_at_nowrap_cpu(const std::shared_ptr<T> ptr_, size_t at, T value) {
+  ptr_.get()[at] = value;
+}
```

Comment:
You're right that this breaks design, but it can also be avoided by just

```c++
template <typename T>
EXPORT_SYMBOL T awkward_index_getitem_at_nowrap(const T* ptr, int64_t offset, int64_t at) {
  return ptr[(size_t)(offset + at)];
}

template <typename T>
EXPORT_SYMBOL void awkward_index_setitem_at_nowrap(T* ptr, int64_t offset, int64_t at, T value) {
  ptr[(size_t)(offset + at)] = value;
}
```

and used (in Index.cpp) as

```c++
template <typename T>
T IndexOf<T>::getitem_at_nowrap(int64_t at) {
  return awkward_index_getitem_at_nowrap(ptr_.get(), offset_, at);
}

template <typename T>
void IndexOf<T>::setitem_at_nowrap(int64_t at, T value) {
  awkward_index_setitem_at_nowrap(ptr_.get(), offset_, at, value);
}
```

Now the only type that gets passed over the border is a raw pointer, which is fully compatible with the extern C interface. That's what all the other kernel functions do. Most C++ style guides discourage this because the raw pointer doesn't track ownership—if the layer 3 code (libawkward) deletes a buffer before the layer 4 code (cpu-kernels) is done with it, there would be data corruption or a segmentation fault. So the rule that we shouldn't do that is explicit in the design: layer 3 code should ensure the existence of buffers while layer 4 code runs. Since we don't do any multithreading in Awkward Array (that's up to code that _uses_ Awkward Array), this is easy: when a layer 3 function calls a layer 4 function, its thread waits until the layer 4 function is done, so the lifetimes of these buffers are guaranteed.

Some conventions:

   * The names of all kernel functions start with "awkward_" and usually include the class they're closely associated with, such as "awkward_index_" here. All the template instantiations (for `Index8`, `IndexU8`, ..., `Index64`) can then be named `awkward_index8_getitem_at_nowrap`, etc. Unmangled names have no C++ namespaces, so we use "awkward_" as a prefix to avoid collisions (a common practice in C).
   * The function shouldn't be designated "_cpu" because all the GPU functions will be in a separate library. Maybe they'll all have to have distinct names because they live in the same namespace—I haven't figured out how that's going to work yet—but those names will be defined systematically. (Maybe we'll rename all the "awkward_" to "awkward_cpu_", but we'll come to that later.)
   * Also a naming convention: trailing underscores (e.g. `ptr_` vs `ptr`) are for class member data. Once a variable is in a kernel function, there are no classes and therefore no class member data.
   * All integer types in the interface have fully specified bit widths. Even though `at` will be used as an array position, `size_t`, it's passed as an `int64_t` and is cast inside the kernel function. C's architecture-dependent integer widths are horrible, so we rein in the madness by making `int64_t` the basic integer type, delving into the others only when necessary. In CI, the Windows 32-bit build (where `size_t` is 32-bit, rather than 64-bit) catches and warns us about implicit integer type conversions, so we make all the integer type conversions explicit. Yes, we're not protected against integer overflows, but at least we know where the overflows can happen because there's an explicit conversion in the code. The unsigned integer types are the worst: reasonable-looking subtractions and checks for `i >= 0` can behave unexpectedly if the integer `i` is declared as unsigned, as is the case with `size_t`. After encountering several bugs related to this (infinite loops and crashes), I've developed a policy of converting to `size_t` as late as possible.
   * The `offset` accounts for the fact that `ptr` refers to the base pointer of each buffer, but a particular Index might be a view into a subset of that buffer that doesn't start at zero. For example, perhaps the user sliced an array with a slice that doesn't start at zero: `array[10:15]`. The slice (`array[10:15]`) shares data with the original array (`array`) to avoid an unnecessary copy. However, both objects have to increment the same reference counter, so their `ptr_` shared pointers point at the same base address (`&array[0]`). They are distinguished by having a different offset (where it begins) and length (how far it goes; equivalently, we could have stored `start = offset` and `stop = offset + length`, but having the `length` as an explicit variable is more useful).
   * These two kernels are extremely unusual in that they don't return an `Error` struct and update arguments in place for their regular outputs. 99% of the kernels have this signature so that error-handling can be done in a systematic way through an interface that doesn't support exceptions. (This is a common pattern for C and Go, languages without try-catch logic.) However, I think these two functions are special enough to join the 1% that break this pattern, along with `awkward_regularize_rangeslice` (which is an exception because it's the only function that's also called by other kernels).

Reply:
Yup, thanks for the detailed explanation. I understood everything. As far as error checking goes, as far as the new commit goes, we can make provision for checking overflows in the `get` and `set` methods.

Reply:
I haven't made any decision on overflow checking yet, so for now we don't check overflows. It's a trade-off: safety is some extreme corner cases vs speed everywhere. Usually the right thing to do is to use a numeric type that's "much bigger" than any number you're likely to encounter, and `int64_t` ought to be that, for most purposes.

If all of the bullet points in this have been addressed (I haven't checked, but I'll do another detailed scan before the merge), then you can resolve this conversation to reduce the noise in the comment thread.

Reply:
Yeah, I think I have addressed the rest of the stuff.